### ISSUE TITLE
Integrating with the test runner

### ISSUE THREAD

**Issue by nicojs** [stryker-mutator/stryker-net#183]
By far the next biggest performance improvement is integrating with the test runner. Right now, we support the `dotnet test` command. This spins up a new dotnet process. There is a lot of overhead here before dotnet even starts executing the tests:

* OS Allocation of process resources
* dotnet allocation of heap + other overhead
* Loading of all assemblies from disk
* Exiting of the process.

It would be great if we can host the child process ourselves. In this process we can load everything once and reuse that process to run tests (each time with a different ACTIVE_MUTATION environment variable). Having the overhead just once will dramatically improve performance. 

In Stryker (for JS) this is done with a test runner plugin system. The command test runner is just one of the available implementations. Other examples are: `MochaTestRunner` (stryker-mocha-runner) and the `JestTestRunner` (stryker-jest-runner). Something similar here would be great to have. Not sure if it needs to be it's own separate dependency.

If you open a solution in Visual Studio, you can see the tests on the left. It doesn't matter if they are implemented in MSTest, NUnit or XUnit. This involves some kind of TestAdapter interface. Maybe we can leverage this api as well, so we only need to implement it once, instead of for each test runner.

**Comment by Mobrockers** [comment]
Useful links?:
https://github.com/Microsoft/vstest
https://github.com/JetBrains/TeamCity.VSTest.TestAdapter/wiki

**Comment by simondel** [comment]
I've removed the hacktoberfest label for now because this will take a while to research, review and develop.

**Comment by nicojs** [comment]
So it should get 10 hacktoberfest labels, right? 😎

**Comment by richardwerkman** [comment]
During my internship I also experimented with several testrunners. Both XUnit and NUnit have nuget packages available to run test projects from code. I couldn't get the XUnit one to run from a .NET Standard 2.0 project. But the NUnit runner seemed to work perfectly and super quick.

It offers the following advantages:
- I had to pass the runner an in memory assembly, so we can run the tests all in memory. (a rerun would be super fast)
- It offers hooks like:
  - TestStarted
  - TestFinished (with result)
- Test exporation can be configured (filtering)

This would speed up Stryker.NET enormously, but of course only on NUnit test projects... 

Would implementing a custom NUnit testrunner inside Stryker.NET be worth the work, or do we want to focus on the VSTest adapter only?

**Comment by Mobrockers** [comment]
All test frameworks that want visual studio support need to implement the vstest testadapter for test discovery. This would most likely also include for example specflow as well as unit test frameworks.

**Comment by dupdob** [comment]
Is anyone looking into this? I would be interested to contribute as this would ultimately reduce run time by order of magnitudes thanks to:
- reducing overhead
- marking mutants as killed on first failed test (no need to go through all tests)
- use mutant coverage to optimise further

**Comment by dupdob** [comment]
This is more or less what I had in mind. I am contemplating working on a demonstrator using the nunit portable agent. This would require backporting it to vstest later on.
Here is a list of intermediate step I could work on:

1. Generate mutant coverage (which mutants is actually executed by any test). This could provide useful metrics as of now. I guess the difficult part would be extracting the statistics from the run process
2. Offer to skip test runs for non covered mutants (hence assumed as survivors) 
       ==> Improve performance for projects with low test coverage.
3. Identify tests that do not cover any mutant, offer to skip them from each run.
       ==>  Could improve performance for projects with many tests
4. Optimise executed tests according to mutant coverage (i.e. run only the tests that cover the 'current' mutant)
     ==> Should dramatically improve performance

Step 1 requires to able to access coverage results from Stryker
Step 2 looks simple
Step 3 requires to control which tests are run
Step 4 require analysis of coverage data to build strategy

None of those steps require using an in process test runner from a theorical standpoint, but may prove too difficult to achieve without one. That was this research is about

**Comment by Mobrockers** [comment]
My plan was as a next step after getting vstest integration working to investigate using a custom data collector to hook into the vstest process. See: https://github.com/Microsoft/vstest/blob/master/test/TestAssets/OutOfProcDataCollector/SampleDataCollector.cs

Using this we might be able to solve all the above problems. This could be researched separate from vstest integration in theory.

**Comment by richardwerkman** [comment]
The DataCollector looks very promising indeed! If we may believe [their documentation](https://github.com/Microsoft/vstest-docs/blob/master/RFCs/0006-DataCollection-Protocol.md) it allows us to execute some code between each individual testcase. If we could insert a line in each mutant like:
``` csharp
File.Write("C://RootDir/MyProject/MutantMapping.log", $"{mutantId}");
```
And after each testcase read that file. We could know what mutants were active during that testcase. We could maybe use the Environment too for this? If we can access the test process Environment from within the DataCollector of course.

**Comment by dupdob** [comment]
I would fear impact on performance. I am pretty sure writing to a file on each invocation would slow some tests to a crawl

**Comment by Mobrockers** [comment]
We can collect the result in-memory and write once only at the end of the testrun. We need to do this unless we can figure out another way to get the data out of the datacollector and into stryker.

**Comment by richardwerkman** [comment]
Note that this only has to happen during the initial testrun. All testruns after that can skip this since the mapping is already known then. The initial testrun would look a bit like this:

1. Startup vstest
2. Start testrun with Custom DataCollector
3. TestCase starts running
4. All mutants that are hit write their 'id' to the file
5. DataCollector hits 'TestCaseEndEvent' and reads all Mutant id's from the file. Then emties the file
6. Return to 3 for next TestCase unless no more testcases
7. DataCollector returns complete mapping of all mutants and testcases to the testhost.
8. Switch a flag so that further testruns will not collect any more data

This would indeed have great impact on the initial testrun. But the mapping will have much greater positive impact on Stykers performance.

Also like @Mobrockers said we could try to keep the mutant 'ids' in memory unit the end of the testcase and then write them once.

**Comment by nicojs** [comment]
> This would indeed have great impact on the initial testrun. But the mapping will have much greater positive impact on Stykers performance.

![https://i.imgflip.com/1vbelm.jpg](https://i.imgflip.com/1vbelm.jpg)

This is especially true for mutation testing. The best thing you can do is create a couple of performance tests and run them each night (small, medium, large project). That's the only way you know for sure something has "great impact"

**Comment by dupdob** [comment]
I will of course execute performance related tests. I was expressing a fear regarding performance.
I think that integration strategy will face several lifecycle/timing related constraints that will drive the actual implementation. But we are not there yet.

For the time being, my focus is on updating Stryker workflow to execute a test run with mutated code, no active mutant and coverage activated, i.e. what I identified as step (1) 😉.

**Comment by dupdob** [comment]
I will look into namedpipe to create an IPC channel between stryker and dotnest test. it will allow real time interaction, better than report parsing after a test pass

**Comment by Mobrockers** [comment]
PR to add vstest.console.dll support to vstest client library has been submitted: https://github.com/Microsoft/vstest/pull/1893

The change worked on my system to start vstest using dotnet from stryker and tests were discovered and run by stryker through vstest.

**Comment by dupdob** [comment]
Some update: I have namedpipes running smoothly and stable, after a few fights.
I have reached maturity level 2: (my) Stryker now assumes non-covered mutants are survivors hence do not test them.

Code available here:
https://github.com/dupdob/stryker-net/tree/mutant_coverage

**Comment by Mobrockers** [comment]
Implemented by #319

Next steps can be picked up in #388
